### PR TITLE
Load exception class via JNI_OnLoad

### PR DESCRIPTION
**Pull request by wolfs**
so all threads can access it and don't fall back to
the system classloader.

### CODE REVIEW

File: src/file-events/cpp/generic_fsnotifier.cpp [hunk 1]
```diff
@@ -306,3 +302,21 @@ Java_net_rubygrapefruit_platform_internal_jni_AbstractFileEventFunctions_00024Na
         rethrowAsJavaException(env, e);
     }
 }
+
+jclass JniConstants::nativeExceptionClass;
```
[the remark this answers is not in the document]
Comment (wolfs) [reply src/file-events/cpp/generic_fsnotifier.cpp]: I got the inspiration from here: https://github.com/guardianproject/IOCipher/blob/master/jni/JniConstants.h

[the remark this answers is not in the document]
Comment (lptr) [reply src/file-events/cpp/generic_fsnotifier.cpp]: I don't see similar redundant declarations of the fields in IOCipher. Maybe I missed something, but why is this necessary?

File: src/file-events/cpp/generic_fsnotifier.cpp [hunk 2]
```diff
@@ -306,3 +302,21 @@ Java_net_rubygrapefruit_platform_internal_jni_AbstractFileEventFunctions_00024Na
...
+    jclass result = reinterpret_cast<jclass>(env->NewGlobalRef(exceptionClass));
+    nativeExceptionClass = result;
+    env->DeleteLocalRef(exceptionClass);
+}
+
+extern "C" jint JNICALL JNI_OnLoad(JavaVM* vm, void*) {
+    JNIEnv* env;
+    if (vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6) != JNI_OK) {
+        return -1;
+    }
```
[the remark this answers is not in the document]
Comment (wolfs) [reply src/file-events/cpp/generic_fsnotifier.cpp]: Done.

File: src/file-events/cpp/generic_fsnotifier.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply src/file-events/cpp/generic_fsnotifier.cpp]: Using a static variable now.

File: src/file-events/cpp/generic_fsnotifier.cpp [hunk 3]
```diff
@@ -306,3 +302,21 @@ Java_net_rubygrapefruit_platform_internal_jni_AbstractFileEventFunctions_00024Na
         rethrowAsJavaException(env, e);
     }
 }
+
+jclass JniConstants::nativeExceptionClass;
+
+void JniConstants::init(JNIEnv* env) {
+    jclass exceptionClass = env->FindClass("net/rubygrapefruit/platform/NativeException");
+    jclass result = reinterpret_cast<jclass>(env->NewGlobalRef(exceptionClass));
```
[the remark this answers is not in the document]
Comment (wolfs) [reply src/file-events/cpp/generic_fsnotifier.cpp]: Done.

File: src/file-events/cpp/generic_fsnotifier.cpp [hunk 4]
```diff
@@ -306,3 +302,21 @@ Java_net_rubygrapefruit_platform_internal_jni_AbstractFileEventFunctions_00024Na
...
+    nativeExceptionClass = result;
+    env->DeleteLocalRef(exceptionClass);
+}
+
+extern "C" jint JNICALL JNI_OnLoad(JavaVM* vm, void*) {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply src/file-events/cpp/generic_fsnotifier.cpp]: Applied your suggestion.

File: src/file-events/cpp/generic_fsnotifier.cpp [hunk 5]
```diff
@@ -306,3 +302,21 @@ Java_net_rubygrapefruit_platform_internal_jni_AbstractFileEventFunctions_00024Na
...
     }
 }
+
+jclass JniConstants::nativeExceptionClass;
+
+void JniConstants::init(JNIEnv* env) {
+    jclass exceptionClass = env->FindClass("net/rubygrapefruit/platform/NativeException");
+    jclass result = reinterpret_cast<jclass>(env->NewGlobalRef(exceptionClass));
+    nativeExceptionClass = result;
+    env->DeleteLocalRef(exceptionClass);
```
[the remark this answers is not in the document]
Comment (wolfs) [reply src/file-events/cpp/generic_fsnotifier.cpp]: 👍

**Comment by wolfs** [comment]
@lptr I addressed your comments, PTAL!

File: src/file-events/headers/generic_fsnotifier.h [hunk 6]
```diff
@@ -178,6 +178,17 @@ class TerminateCommand : public Command {
...
+struct JniConstants {
+    JniConstants(JNIEnv* env);
+    void unload(JNIEnv* env);
+
+    const jclass nativeExceptionClass;
+    const jclass classClass;
+    const jclass nativeFileWatcherClass;
+};
+
+extern JniConstants* jniConstants;
```
Comment (lptr) [inline src/file-events/headers/generic_fsnotifier.h]: Why `extern` and not `static`? We don't want this to be accessible outside, no?

Comment (wolfs) [reply src/file-events/headers/generic_fsnotifier.h]: If we declare it `static` in the header, then every object file will have the field compiled in. I don't think that is what we want. We can remove the declaration from the header completely, given that it is currently only used in `generic_fsnotifier.cpp`. If we want other `.cpp` files to use the `jniConstants`, then we need to use extern.

Comment (lptr) [reply src/file-events/headers/generic_fsnotifier.h]: OK. I'd remove it from the header until we actually need to export it, but I'm okay with this too.

Comment (lptr) [reply src/file-events/headers/generic_fsnotifier.h]: Actually, I might need these from `logging.cpp`, so let's keep it `extern`.

**Pull request merged**